### PR TITLE
EE-2386 - Populated Submission object

### PR DESCRIPTION
**Pull request by M-casado**
## Ticket reference
[EE-2386](https://www.ebi.ac.uk/panda/jira/browse/EE-2386)

## Overall changes
- [ ] Populated submission object JSON schema
- [ ] Modified other JSON schemas for objects to reference submissions
- [ ] Added submission JSON document example
- [ ] Modified some existing JSON examples to reference submissions

## Future TO-DOs
- [ ] Possibly adding the constraint that every object needs to at least point to a submission project. Waiting for Sabela/Oscar to answer to my inquiry.

### CODE REVIEW

File: examples/json_validation_tests/submission_valid-1.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        "alias" : "submission_XF40", 
+        "center_name" : "EBI-TEST"
+    },
+    "object_title": "Submission XF40",
+    "object_description": "EBI submission project XF40 of 1000 samples and its 1500 sequencing runs",
+    "additional_collaborators": [
+        {
+            "collaborator_rights": "read_and_write",
+            "collaborator_contact_details": {
+                "institution_name": "European Bioinformatics Institute (EBI)",
```
Comment (malloryfreeberg) [inline examples/json_validation_tests/submission_valid-1.json:12]:
```suggestion
                "institution_name": "EMBL European Bioinformatics Institute (EBI)",
```

File: schemas/EGA.DAC.json [hunk 2]
```diff
@@ -38,13 +38,15 @@
         "type": "string",
         "title": "Title of the DAC",
         "description": "Short free-form text that can be used to call out DAC records in searches or displays.",
+        "minLength": 1,
```
Comment (malloryfreeberg) [inline schemas/EGA.DAC.json:41]: What does this change do? Is it related to the addition of the submission object?

File: schemas/EGA.submission.json [hunk 3]
```diff
@@ -4,10 +4,188 @@
     "type": "object",
     "title": "EGA submission metadata schema",
     "meta:version": "0.0.0",
-    "description": "Empty for the moment.",
-    "required": [""],
+    "description": "Metadata schema used by the European Genome-phenome Archive (EGA) to validate its Submission (also known as submission project) metadata object. This object is intended to be an object that others can reference, grouping them by submission details. An EGA user can submit metadata comprising thousands of objects, or just a few that suffice the bare minimum of a comprehensive submission: at least one dataset, with its proper links to other objects. Once a submission is finished, EGA's HelpDesk team will review it and, possibly, accept it. A submission JSON document hold little details, mainly the basic descriptive fields and collaborators array, but its main use is for other objects to reference it. Further details can be found in the EGA-metadata-schema GitHub repository (https://github.com/EbiEga/ega-metadata-schema/tree/main/schemas) and EGA-archive website (https://ega-archive.org/submission/data_access_committee)",
```
[the remark this answers is not in the document]
Comment (malloryfreeberg) [reply schemas/EGA.submission.json:7]: I suggested to remove some bits that are likely to change in the future, to ensure that this schema is sustainable and that we don't have to remember to change it when our processes change.

File: schemas/EGA.submission.json [hunk 4]
```diff
@@ -4,10 +4,188 @@
...
+                "read_only": "Collaborator will only be capable of reading the details of the submission.",
+                "read_and_write": "Collaborator will be able to not only read, but modify or add changes to the submission."
+              }
+            },
+            "collaborator_contact_details": {
+              "$ref": "https://github.com/EbiEga/ega-metadata-schema/tree/main/schemas/EGA.common-definitions.json#/definitions/contact_details"
+            }
+          }
+        }
+      },
```
Comment (malloryfreeberg) [inline schemas/EGA.submission.json:81]: How does this information align with the EBI's current submission process, whereby a submission account (i.e. ega-box-###) is opened for all named individuals in the submission statement? How does this information align with CRG's proposed approach for managing submission accounts?

I'm assuming the information provided by a submitter here doesn't have any actual effect on who can read from / write to a submission account. Could this be confusing to users?

File: examples/json_validation_tests/submission_valid-1.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                "email_address": "example@gmail.com"
+            }
+        },
+        {
+            "collaborator_rights": "read_only",
+            "collaborator_contact_details": {
+                "individual_full_name": "Isaac Asimov",
+                "institution_name": "Boston University",
+                "email_address": "example@gmail.com"
```
Comment (malloryfreeberg) [inline examples/json_validation_tests/submission_valid-1.json:21]: I suggest not using gmail as an example since we do not accept gmail email accounts for users.

**Comment by malloryfreeberg** [comment]
Great job updating all the examples, as well 👍  It was useful to have only the examples and schemas to look at.

I wasn't sure what all the additions of `"minLength": 1,` were for, as they don't appear to be linked to adding the submission object.

File: schemas/EGA.DAC.json (hunk 2 above)
Comment (M-casado) [reply schemas/EGA.DAC.json:41]: It basically means that the given string needs to be of at least length 1. In other words, no empty string is allowed. 
As you mentioned it's not related to submission objects directly, just generally.

File: schemas/EGA.submission.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (M-casado) [reply schemas/EGA.submission.json:7]: Okay, that's sensible. I was just trying to add more context as to why we would have another grouping object. It sounds difficult to understand unless you mention that HD would review batches of "submission" collections.

File: schemas/EGA.submission.json (hunk 4 above)
Comment (M-casado) [reply schemas/EGA.submission.json:81]: That is a really good question. The conflict was to have a JSON Schema we can share metadata between institutions with, but also a JSON Schema that users could use to provide said metadata. 

Based on the demos we saw of the Submitter Portal, users can share their submissions and collaborate. CRG members also agreed to have this information within the JSON documents. **Although we did not mention explicitly that what users put into this node would govern who can collaborate with which submission, that was what I understood**. 

If, instead, it's preferable to have this information provided in another way, we could remove it from the submission JSON schema. We can also create an extended JSON schema with that additional node for EBI-CRG use only. In any case, I would advise to keep this node and also effectively reading collaborators from it.

Regarding how it aligns with EBI's / CRG: the way we created this object was **mainly having CRG's point of view of a submission**. Since these schemas are expected to be use in the long-run, and we will eventually reproduce/reuse this submission concept, I disregarded the current submission concept (perhaps ADD/MODIFY actions could come in handy, though, but that would be for the object-set, not for this concept of submission). I imagine that, either at some point we adapt submission accounts at EBI in a similar fashion to CRG (after all we'll be using the same website), or we can have collaborators being added individually by their email addresses, apart from submission accounts.

Let me know if that answers your question or not, happy to discuss if needed.

File: examples/json_validation_tests/submission_valid-1.json (hunk 5 above)
Comment (M-casado) [reply examples/json_validation_tests/submission_valid-1.json:21]: True

File: schemas/EGA.submission.json [hunk 6]
```diff
@@ -4,10 +4,188 @@
...
+        "type": "string",
+        "title": "Description of the submissions project",
+        "description": "An in-depth description of the submission, including its overall purpose or nature of studies it governs.",
+        "minLength": 1,
+        "examples": [ "EBI submission project XF40 of 1000 samples and its 1500 sequencing runs" ]
+      },
+
+      "additional_collaborators": {
+        "type": "array",
+        "title": "Submission ownership details",
```
Comment (malloryfreeberg) [inline schemas/EGA.submission.json:55]:
```suggestion
        "title": "Submission collaborator details",
```

Based on the attribute description, I think this field is meant to capture individuals who are **not** the data owner / controller / DAC, but can still collaborate on the submission. If so, I suggest changing the attribute title.

File: schemas/EGA.submission.json (hunk 4 above)
Comment (malloryfreeberg) [reply schemas/EGA.submission.json:81]: It's a good idea to encode this information somewhere, and the submission JSON schema seems a good choice. What I don't think we can do is allow permissions to be set by someone providing a submission JSON document. This approach is effectively letting anyone who has access to a submission also control who else can access the submission (which might contain private info) without authorisation. It is only the Data Controller / DAC who can authorise collaborators, and this all has to be done via the DAC Portal or via Helpdesk.

I'm assuming because the name of the attribute is `additional_collaborators` that the data controllers and/or DACs are **not** meant to be listed in this block? 

tl;dr: The `additional_collaborators` block can be used to represent additional (or all?) collaborators, but changes to collaborator information must always go through the DAC Portal or Helpdesk.

Comment (M-casado) [reply schemas/EGA.submission.json]: That's true, and I started to think on whether this may be too confusing for the users: having the node there but not being able to use it. 

Perhaps "submissions" could be submitted through another pathway (i.e. through DAC Portal). I'm not familiar with the DAC Portal, but as far as I remember it has no programmatic interaction, does it? I may be wrong, but during the Submitter Portal demo I think they showed how collaborators could be added directly without going through the DAC Portal. I need to check the recording.

In any case, we could always display in bold that the collaboration property is just for internal usage, and that it won't propagate if it's submitted directly. 

> I'm assuming because the name of the attribute is additional_collaborators that the data controllers and/or DACs are not meant to be listed in this block?

Having the real "owners" of the "submission" there was an initial approach, but I removed them, since the owners of an object are omitted in the other objects. But perhaps these details would come in handy when institutions exchange information. I wouldn't mix too much what users can submit and what institutions exchange as extras, since it leads to, for instance, the current dilemma of collaborators.

**Review by malloryfreeberg — APPROVED** [review]
